### ISSUE TITLE
CGDataProvider and SIGSEGV

### ISSUE THREAD

**Issue by derekdreery** [servo/core-foundation-rs#431]
It seems as though `CGDataProvder::from_buffer` segfaults if the buffer is empty. A reproducing test case would be something like

```rust
let buf: Arc<[u8]> = Arc::new([]);
CGDataProvider::from_buffer(buf)
```

I don't have a macOS box for testing, so this is based on experimentation with github CI. Apologies if I'm mistaken about this issue.

**Comment by jrmuizel** [comment]
What version of the core-graphics crate are you using?

**Comment by jrmuizel** [comment]
Also, 
```rust
let buf: Arc<[u8]> = Arc::new([]);
CGDataProvider::from_buffer(buf)
```
gives me:
```
error[E0277]: the size for values of type `[u8]` cannot be known at compilation time
  --> src/main.rs:7:33
   |
7  |     CGDataProvider::from_buffer(buf);
   |                                 ^^^ doesn't have a size known at compile-time
   | 
  ::: /Users/jrmuizel/.cargo/registry/src/github.com-1ecc6299db9ec823/core-graphics-0.22.2/src/data_provider.rs:53:24
   |
53 |     pub fn from_buffer<T: AsRef<[u8]> + Sync + Send>(buffer: Arc<T>) -> Self {
   |                        - required by this bound in `core_graphics::data_provider::CGDataProvider::from_buffer`
   |
   = help: the trait `Sized` is not implemented for `[u8]`
```

**Comment by jrmuizel** [comment]
A variant with a zero len Vec also crashes in ERROR_CGDataProviderCreateWithData_BufferIsNotReadable. We could potentially panic earlier.

**Comment by derekdreery** [comment]
So the place this came up is in the [`piet` project](https://github.com/linebender/piet/blob/master/piet-coregraphics/src/lib.rs#L274), where creating image with zero width and height.

**Comment by jrmuizel** [comment]
So yeah, you only get to have a 0 length CGDataProvider if it's pointing at something dereferenceable. You could substitute at 1 length buffer in piet to work around it.

**Comment by derekdreery** [comment]
Thanks for the advice. The worry is that you can corrupt memory with a safe method. Did you way that you didn't get a segfault: you got a panic instead?

**Comment by jrmuizel** [comment]
No, I got a segfault as well, but it's a segfault specifically designed for catching unreadable buffers and so there doesn't seem to be a danger in corrupting memory.

I'm proposing adding a panic CGDataProvider::from_buffer if the length is 0 so that we don't get the segfault at all.

**Comment by derekdreery** [comment]
All sounds good :)

**Comment by derekdreery** [comment]
Could you also add a `+ ?Sized` to the constraints on `T` in `DataProvider::from_buffer`? That way you can use a `Arc<[u8]>`, which currently isn't possible because `[u8]` is unsized. I checked and the code will still be correct.
